### PR TITLE
J2472

### PR DESCRIPTION
**Pull request by rozza**
Theres a couple of preliminary commits which came about during the writing the documentation.

A preview is available: http://rosslawley.co.uk/mongo-java-driver/3.5/bson/pojos

### CODE REVIEW

File: docs/reference/content/bson/pojos.md [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+
+The 3.5 release of the driver adds POJO support via the [`PojoCodec`]({{<apiref "org/bson/codecs/pojo/PojoCodec.html">}}), which allows for 
+direct serialization of POJOs to and from BSON. Internally, each `PojoCodec` utilizes a
+[`ClassModel`]({{<apiref "org/bson/codecs/pojo/ClassModel.html">}}) instance to store metadata about how the POJO should be serialized.
+
+A `ClassModel` for a POJO includes:
+
+  * The class of the POJO.
+  * A new instance factory. Handling the creation of new instances of the POJO. By default it requires the POJO to have an empty constructor.
+  * Field information, a list of [`FieldModel`]({{<apiref "org/bson/codecs/pojo/FieldModel.html">}}) instances that contain all the field metadata.
```
Comment (jyemin) [inline docs/reference/content/bson/pojos.md]: The doc should state what type of fields are serialized by default: all non-transient, non-final fields, I believe

File: docs/reference/content/bson/pojos.md [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+  * An optional IdField. By default the `_id` or `id` field in the POJO.
+  * Type data for the POJO and it's fields to work around type erasure.
+  * An optional discriminator value. The discriminator is the value used to represent the POJO class being stored and by default is the
```
Comment (jyemin) [inline docs/reference/content/bson/pojos.md]: This is confusing.  It says the discriminator is optional, then that by default is the simple class name.  IIRC by default there is no discriminator.

File: docs/reference/content/bson/pojos.md [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+    simple name of the POJO class.
+  * An optional discriminator key. The document field name for the discriminator which has the default of `_t`.
+  * The use discriminator flag.  This determines if the discriminator should be serialized. By default it is off.
+  
+Each `FieldModel` includes:
+
+  * The field name.
+  * The document field name, which is the key for the value when serialized to BSON. By default it is the same as the field name.
+  * Type data, to work around type erasure.
+  * An optional `Codec` for the field. Which allows for fine grained control over how the field is encoded and decoded.
```
Comment (jyemin) [inline docs/reference/content/bson/pojos.md]: Grammar: Don't start a sentence with "Which"

File: docs/reference/content/bson/pojos.md [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+  * A serialization checker. This checks if the value should be serialized. By default, `null` values are not serialized.
+  * Use discriminator flag which is only used when serializing other POJOs. By default it is off.
```
Comment (jyemin) [inline docs/reference/content/bson/pojos.md]: Clarify how this interacts the the discriminator value in the class model.

File: docs/reference/content/bson/pojos.md [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+
+The instance of `new Person("Ada", "Lovelace");` would be serialized to the equivalent of `{ firstName: "Ada", lastName: "Lovelace"}`. 
+
+Notice the `address` field is omitted because it hasn't been set and has a `null` value. If the person instance contained an address, it 
+would be stored as a sub document and use the `CodecRegistry` to look up the `Codec` for the `Address` class and use that to 
+encode and decode the address value.
+
+### Generics support
+
+Generics are fully supported, during the creation of a `ClassModelBuilder` the type parameters are inspected and saved to work around type
```
Comment (jyemin) [inline docs/reference/content/bson/pojos.md]: Grammar: run-on sentence

File: docs/reference/content/bson/pojos.md [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+discriminator.
+
+
+## Advanced configuration
+
+For most scenarios there is no need for further configuration. However, there are some scenarios where custom configuration is required.
+
+### Fields with abstract or interface types.
+
+If a POJO contains a field that has an abstract type or has an interface as it's type, then a discriminator is required. The type and all
```
Comment (jyemin) [inline docs/reference/content/bson/pojos.md]: it's => its

File: docs/reference/content/bson/pojos.md (hunk 1 above)
Comment (rozza) [inline docs/reference/content/bson/pojos.md]: Done.

File: docs/reference/content/bson/pojos.md [hunk 7]
```diff
@@ -0,0 +1,262 @@
...
+The 3.5 release of the driver adds POJO support via the [`PojoCodec`]({{<apiref "org/bson/codecs/pojo/PojoCodec.html">}}), which allows for 
+direct serialization of POJOs to and from BSON. Internally, each `PojoCodec` utilizes a
+[`ClassModel`]({{<apiref "org/bson/codecs/pojo/ClassModel.html">}}) instance to store metadata about how the POJO should be serialized.
+
+A `ClassModel` for a POJO includes:
+
+  * The class of the POJO.
+  * A new instance factory. Handling the creation of new instances of the POJO. By default it requires the POJO to have an empty constructor.
+  * Field information, a list of [`FieldModel`]({{<apiref "org/bson/codecs/pojo/FieldModel.html">}}) instances that contain all the field metadata.
+  * An optional IdField. By default the `_id` or `id` field in the POJO.
```
Comment (rozza) [inline docs/reference/content/bson/pojos.md]: Done

File: docs/reference/content/bson/pojos.md (hunk 2 above)
Comment (rozza) [inline docs/reference/content/bson/pojos.md]: Done. The default convention sets it - but thats documented below.

File: docs/reference/content/bson/pojos.md (hunk 3 above)
Comment (rozza) [inline docs/reference/content/bson/pojos.md]: Done.

File: docs/reference/content/bson/pojos.md (hunk 4 above)
Comment (rozza) [inline docs/reference/content/bson/pojos.md]: Done

File: docs/reference/content/bson/pojos.md (hunk 5 above)
Comment (rozza) [inline docs/reference/content/bson/pojos.md]: Split up.

File: docs/reference/content/bson/pojos.md (hunk 6 above)
Comment (rozza) [inline docs/reference/content/bson/pojos.md]: Done.

**Comment by rozza** [comment]
Agreed, was thinking about adding a `PojoQuickTour` as well.

File: docs/reference/content/bson/pojos.md [hunk 8]
```diff
@@ -17,23 +17,24 @@ A `ClassModel` for a POJO includes:
 
   * The class of the POJO.
   * A new instance factory. Handling the creation of new instances of the POJO. By default it requires the POJO to have an empty constructor.
-  * Field information, a list of [`FieldModel`]({{<apiref "org/bson/codecs/pojo/FieldModel.html">}}) instances that contain all the field metadata.
+  * Field information, a list of [`FieldModel`]({{<apiref "org/bson/codecs/pojo/FieldModel.html">}}) instances that contain all the field metadata. By default all non static and non transient fields.
```
Comment (jyemin) [inline docs/reference/content/bson/pojos.md]: "By default this includes all non-static and non-transient fields.

Comment (rozza) [inline docs/reference/content/bson/pojos.md]: Done.

**Comment by rozza** [comment]
Manually rebased and merged

**Pull request closed without merging**